libraries/chain/steem_evaluator.cpp

```diff
@@ -2342,89 +2342,134 @@ namespace golos { namespace chain {
-                auto regenerated_power = (STEEMIT_100_PERCENT * elapsed_seconds) / STEEMIT_VOTE_REGENERATION_SECONDS;
-                auto current_power = std::min<int64_t>(delegator.voting_power + regenerated_power, STEEMIT_100_PERCENT);
-                auto max_allowed = delegator.vesting_shares * current_power / STEEMIT_100_PERCENT;
-                GOLOS_CHECK_LOGIC(delegated + delta <= max_allowed,
-                    logic_exception::delegation_limited_by_voting_power,
-                    "Account allowed to delegate a maximum of ${v} with current voting power = ${p}",
-                    ("v",max_allowed)("p",current_power)("delegated",delegated)("delta",delta));
-
-                if (!delegation) {
-                    GOLOS_CHECK_OP_PARAM(op, vesting_shares, {
-                        GOLOS_CHECK_LOGIC(op.vesting_shares >= min_delegation,
-                            logic_exception::cannot_delegate_below_minimum,
-                            "Account must delegate a minimum of ${v}",
-                            ("v",min_delegation)("vesting_shares",op.vesting_shares));
-                    });
-                    _db.create<vesting_delegation_object>([&](vesting_delegation_object& o) {
-                        o.delegator = op.delegator;
-                        o.delegatee = op.delegatee;
-                        o.vesting_shares = op.vesting_shares;
-                        o.min_delegation_time = now;
-                    });
-                }
-                _db.modify(delegator, [&](account_object& a) {
-                    a.delegated_vesting_shares += delta;
-                });
-            } else {
-                GOLOS_CHECK_OP_PARAM(op, vesting_shares, {
-                    GOLOS_CHECK_LOGIC(op.vesting_shares.amount == 0 || op.vesting_shares >= min_delegation,
-                        logic_exception::cannot_delegate_below_minimum,
-                        "Delegation must be removed or leave minimum delegation amount of ${v}",
-                        ("v",min_delegation)("vesting_shares",op.vesting_shares));
-                });
-                _db.create<vesting_delegation_expiration_object>([&](vesting_delegation_expiration_object& o) {
-                    o.delegator = op.delegator;
-                    o.vesting_shares = -delta;
-                    o.expiration = std::max(now + STEEMIT_CASHOUT_WINDOW_SECONDS, delegation->min_delegation_time);
-                });
+        void delegate_vesting_shares_with_interest_evaluator::do_apply(const delegate_vesting_shares_with_interest_operation& op) {
+            ASSERT_REQ_HF(STEEMIT_HARDFORK_0_19__756, "delegate_vesting_shares_with_interest_operation");
+
+            const auto& median_props = _db.get_witness_schedule_object().median_props;
+
+            auto add_fields_to_vdo = [&](vesting_delegation_object& o) {
+                o.interest_rate = std::min(op.interest_rate, median_props.max_delegated_vesting_interest_rate);
```

Comment:
Propose to abort operation with fail if `op.interest_rate` greater than `median_props.max_delegated_vesting_interest_rate`. This is more clear. Furthermore, blocklog will contain effective `op.interest_rate` value with no requirement to calculate whole state to know, is it affected by median_props